dstep/driver/Application.d

```diff
@@ -110,115 +86,84 @@ class Application
+                message.put(diagnostics.format);
+                message.put("\n");
+            }
+        }
+
+        enforce!DStepException(translate, message.data);
     }
 
-    this (
-        const Configuration config,
-        string inputFile,
-        string outputFile)
+    static string makeDefaultOutputFile(string inputFile, bool useBaseName = true)
     {
-        this.config = config;
-        this.inputFile = inputFile;
-        this.outputFile = outputFile;
+        if (useBaseName)
+            return Path.setExtension(Path.baseName(inputFile), "d");
+
+        return Path.setExtension(inputFile, "d");
     }
 
-    void startConversion ()
+    static string[] makeOutputFiles(Configuration config)
     {
-        import std.algorithm.iteration : map;
+        import std.algorithm;
         import std.array;
+        import std.range;
 
-        index = Index(false, false);
-
-        auto include_flags = compiler
-            .extraIncludePaths.map!(e => "-I" ~ e).array();
-        translationUnit = TranslationUnit.parse(
-            index,
-            inputFile,
-            config.clangParams ~ include_flags,
-            compiler.extraHeaders);
-
-        diagnostics = translationUnit.diagnostics;
-
-        enforceCompiled();
+        auto inputFiles = config.inputFiles;
 
-        if (exists(inputFile))
+        // when only one input file is supplied, -o argument is
+        // interpreted as file path, otherwise as base directory path
+        if (inputFiles.length == 1)
         {
-            import std.algorithm : map;
-            import std.array : array;
-
-            Options options = this.config.toOptions(inputFile, outputFile);
-
-            auto translator = new Translator(translationUnit, options);
-            translator.translate;
+            return [config.output.empty
+                ? makeDefaultOutputFile(inputFiles.front, false)
+                : config.output];
         }
-    }
-
-    ~this ()
-    {
-        clean();
-    }
-
-private:
+        else
+        {
+            alias fmap = file => Path.buildPath(
+                config.output,
+                makeDefaultOutputFile(file, false));
 
-    void clean ()
-    {
-        translationUnit.dispose;
-        index.dispose;
+            return inputFiles.map!fmap.array;
+        }
     }
 
-    bool anyErrors ()
+    static TranslationUnit[] makeTranslationUnits(Configuration config)
     {
-        return diagnostics.length > 0;
-    }
+        import std.parallelism;
 
-    void enforceCompiled ()
-    {
-        import std.array : Appender;
-        import std.exception : enforce;
+        Index translationIndex = Index(false, false);
+        Compiler compiler;
 
-        bool translate = true;
-        auto message = Appender!string();
+        auto translationUnits = new TranslationUnit[config.inputFiles.length];
 
-        foreach (diag ; diagnostics)
+        foreach (index, ref unit; taskPool.parallel(translationUnits))
```

Comment:
Can this be replaced with a map?

Reply:
Do you know why this doesn't compile:

```
    import std.parallelism;

    Index translationIndex = Index(false, false);
    Compiler compiler;

    TranslationUnit delegate(string) functor = inputFile => TranslationUnit.parse(
        translationIndex,
        inputFile,
        config.clangParams ~ compiler.extraIncludeFlags,
        compiler.extraHeaders);

    return taskPool.map!functor(config.inputFiles, 1).array;
```

Error:
`dstep/driver/Application.d(162,24): Error: template instance map!(functor) cannot use local 'functor' as parameter to non-global template map(functions...)`

Reply:
?

Reply:
I think this is the reason I haven't used the map in the first place (I created the change quire some time ago, it only the PR which were created yesterday).

https://issues.dlang.org/show_bug.cgi?id=5710

This is (another) stupid limitation of D : /.

Reply:
Won't fix.

Reply:
Ah, ok. That issue again 😞.